src/mgr/ActivePyModules.cc

```diff
@@ -985,6 +1094,161 @@ PyObject* ActivePyModules::get_perf_schema_python(
   return f.get();
 }
 
+PyObject* ActivePyModules::get_perf_schema_python(
+  const std::string &svc_type,
+  const std::string &svc_id)
+{
+  without_gil_t no_gil;
+  std::lock_guard l(lock);
+
+  DaemonStateCollection daemons;
+
+  if (svc_type == "") {
+    daemons = daemon_state.get_all();
+  } else if (svc_id.empty()) {
+    daemons = daemon_state.get_by_service(svc_type);
+  } else {
+    auto key = DaemonKey{svc_type, svc_id};
+    // so that the below can be a loop in all cases
+    auto got = daemon_state.get(key);
+    if (got != nullptr) {
+      daemons[key] = got;
+    }
+  }
+
+  auto f = with_gil(no_gil, [&] {
+    return PyFormatter();
+  });
+
+  auto dump_sub_counter_information = [] (PyFormatter *f, PerfCounterType type){
+    Formatter::ObjectSection counter_section(*f, type.counter_name);
+    f->create_unique("description", type.description);
+    if (!type.nick.empty()) {
+      f->dump_string("nick", type.nick);
+    }
+    f->dump_unsigned("type", type.type);
+    f->dump_unsigned("priority", type.priority);
+    f->dump_unsigned("units", type.unit);
+  };
+
+  auto dump_counter_with_labels = [&dump_sub_counter_information] (PyFormatter *f, auto key_labels, auto type){
+    f->open_object_section(""); // counter should be enclosed by array
+
+    for (Formatter::ObjectSection labels_section{*f, "labels"}; const auto &label: key_labels){
+      f->dump_string(label.first, label.second);
+    }
+
+    f->open_object_section("counters");
+    dump_sub_counter_information(f, type);
+  };
+
+
+  if (!daemons.empty()) {
+    for (auto& [key, state] : daemons) {
+      std::lock_guard l(state->lock);
+      with_gil(no_gil, [&, key=ceph::to_string(key), state=state] {
+        string_view key_name, prev_key_name;
+        vector<pair<string_view,string_view>> prev_key_labels;
+        Formatter::ObjectSection counter_section(f, key.c_str()); // Main Object Section
+        std::optional<Formatter::ArraySection> array_section;
+
+        for (auto ctr_inst_iter : state->perf_counters.instances) {
```

Comment:
structured binding?
i.e.:
for (const auto& [counter_name_with_labels, [[maybe_unsued]] _ ] : state->perf_counters.instances

(the 'maybe_unsued' won't be required in C++23)

Reply:
done!